File: src/DotNetMatrix/GeneralVector.cs
using System;
using System.Linq;
using System.Collections.Generic;

namespace DotNetMatrix
{
	public class GeneralVector
	{
		private RowVector InnerVector;

		public GeneralVector (int Length)
		{
			InnerVector = new RowVector (1, Length);
		}

		public GeneralVector(double [] array)
			: this(array.Length)
		{
			array.Select ((e, i) => this [i] = e); 
		}

		public double Magnitude
		{
			get
			{
				return InnerVector.NormF ();
			}
		}

		public double this[int index]
		{
			get
			{
				return InnerVector [0] [index];
			}

			set
			{
				InnerVector [0] [index] = value;
			}
		}

		public void RescaleToUnitInterval()
		{
			if (InnerVector [0].Max() > 0) {
				for (int i = 0; i < InnerVector [0].Count; i++)
					InnerVector [0] [i] = InnerVector [0] [i] / InnerVector [0].Max();
			}
		}

		public void InitToZero()
		{
			InnerVector [0].Select (e => e = 0);
		}

		public double UnitaryDotProduct(GeneralVector operand)
		{
			double total = 0;
			this.ToList
				.Select((e, i) => total +=
					((e * operand [i]) /
						(this.Magnitude * operand.Magnitude)));
			return total;
		}

		public List<double> ToList
		{
			get
			{
				return InnerVector [0];
			}
		}

		public GeneralMatrix Multiply (GeneralMatrix matrix)
		{
			return InnerVector.Multiply (matrix);
		}
	}
}

File: src/Jama/Jama/Jama/GeneralVector.java
package Jama;

import java.util.ArrayList;
import java.util.Collections;

public class GeneralVector
{
	RowVector InnerVector;
	
	public GeneralVector(int Length)
	{
		InnerVector = new RowVector(1,Length);
	}
	
	public GeneralVector(double [] array)
	{	
		this(array.length);
		for(int i=0; i < array.length; i++)
		{
			InnerVector.set(0, i, array[i]);
		}
	}
	
	public double get(int index)
	{	
		return InnerVector.get(0, index);
	}
	
	
	public void set(int index, double value)
	{
		InnerVector.set(0, index, value);
	}
	
	public void initToZero()
	{
		for(int index=0; index < InnerVector.n; index++)
		{
			set(index, 0d);
		}
	}
	
	public void rescaleToUnitInterval()
	{
		
		double max = 0d;
		for(int index=0; index < InnerVector.n; index++)
		{
			max = (Math.abs(get(index)) > Math.abs(max))
			? get(index)
			: max; 
		}
		
		if(max != 0d)
		{
			for(int index=0; index < InnerVector.n; index++)
			{
				set(index, InnerVector.get(0,  index)/max);
			}
		}
	}
	
	public double getMagnitude()
	{
		return InnerVector.normF();
	}
	
	public double unitaryDotProduct(GeneralVector operand)
	{
		double total = 0d;
		
		for(int index=0; index < InnerVector.n; index++)
		{
			total += get(index) * operand.get(index);
		}
		return total/(getMagnitude()*operand.getMagnitude());
	}

	public ArrayList<Double> toList()
	{
		return InnerVector.get(0);
	}
	
	public Matrix Multiply(Matrix matrix)
	{
		return InnerVector.arrayTimes(matrix);
	}
}	
